File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def help(r: Optional[TreeNode]) -> [int, int]:
            # maxDepth, bestResult
            if not r:
                return [0, 0]
            right, left = help(r.right), help(r.left)
            return [1 + max(right[0], left[0]), max(right[1], left[1], right[0] + left[0])]
        [max_depth, best_result] = help(root)
        return best_result

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return help(root) != -1;
    }
private:
    int help(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int right = help(root->right);
        int left = help(root->left);
        
        if (right == -1 || left == -1) {
            return -1;
        }
        if (right - left > 1 || left - right > 1) {
            return -1;
        }
        return 1 + max(left, right);
    }
};
